scipy/stats/_hypotests.py

```diff
@@ -663,3 +663,272 @@ def somersd(x, y=None):
+            `pooled` param.
+            ``pvalue`` the probability of obtaining a test statistic at least
+            as extreme as the one that was actually observed, assuming that
+            the null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Notes
+    -----
+    Barnard's test is an exact test used in the analysis of contingency
+    tables. It examines the association of two categorical variables and
+    is a more powerful alternative than Fisher's exact test
+    for 2x2 contingency tables.
+    When using Barnard exact test, we can assert three different null
+    hypothesis :
+
+    - :math:`H_0 : p_1 \geq p_2` versus :math:`H_1 : p_1 < p_2`,
+      with `alternative` = "less"
+
+    - :math:`H_0 : p_1 \leq p_2` versus :math:`H_1 : p_1 > p_2`,
+      with `alternative` = "greater"
+
+    - :math:`H_0 : p_1 = p_2` versus :math:`H_1 : p_1 \neq p_2`,
+      with `alternative` = "two-sided" (default one)
+
+    References
+    ----------
+    .. [1] Mehta, Cyrus & Senchaudhuri, Pralay. (2003).
+           "Conditional versus Unconditional Exact Tests for Comparing Two
+           Binomials"
+
+    Examples
+    --------
+    An example use of Barnard's test is presented in [1]_.
+
+        Consider the following example of a vaccine efficacy study
+        (Chan, 1998). In a randomized clinical trial of 30 subjects, 15 were
+        inoculated with a recombinant DNA influenza vaccine and the 15 were
+        inoculated with a placebo. Twelve of the 15 subjects in the placebo
+        group (80%) eventually became infected with influenza whereas for the
+        vaccine group, only 7 of the 15 subjects (47%) became infected. The
+        data are tabulated as a 2 x 2 table::
+
+                Vaccine  Placebo
+            Yes     7        12
+            No      8        3
+                    15        15
+    When working with statistical hypothesis testing, we usually use a
+    threshold probability or a significance level upon which we decide
+    to reject or not the null hypothesis :math:`H_0`. A commonly used
+    significance level is 5%. In this example, we are using the
+    `alternative` parameter with the "less" option, because the vaccine has
+    either no effect (:math:`H_0` hypothesis) or a positive effect
+    (:math:`H_1` hypothesis). Therefore, under the null hypothesis
+    (:math:`H_0`), the probability :math:`p_2` of catching the virus
+    **without** any vaccine, is either equal or lower than :math:`p_1`,
+    the probability of catching the virus **with** the vaccine.
+    To compute the p-value with Banard test, let's call `barnard_exact` as
+    follow :
+
+    >>> import scipy.stats as stats
+    >>> res = stats.barnard_exact([[7, 12], [8, 3]], alternative="less")
+    >>> res.statistic
+    1.894...
```

Comment:
Thanks for your feedback, I am going to work on it then :-)

Reply:
Also there is a sign change in the test. I have seen this example in the literature and it is a positive value. Not sure what is happening here.

Reply:
The sign of `TX` depends on the direction you choose for the definition. Most authors seem use `p2-p1`. In the actual code, `p1-p2` is used for consistency with `fisher_exact`, therefore the sign change.

In fact, we have 4 options (or maybe more):
 - Keep `p1-p2`.
 - Change to `p2-p1`. But in this case if we want to keep the signification of the name `less` and `greater`, we need to inverse them, and we are not consistent with the choice made in `stats.fisher_exact()`.
 - Change to `p2-p1`, and change the indexes selection (lines 897 to 900). With this, we keep the consistency with `stats.fisher_exact()`, but we brake the understandably of the statistics (_pvalue when « H₁ = *less* » means the probability under H₀ than the test stats is *less* than the observed_).
 - Keep as actual, but change the sign on return. (The very dirty solution IMHO).

Reply:
Thanks @jb-leger for the details. Being consistent with Fisher sounds like a good idea. What do you think @mdhaber? If we keep this then just update the test with the correct sign and I would also suggest to add a note in the code as someone coming back to it might question this.

Reply:
Yes, I am going to change the sign :-)

Reply:
Just to be clear @tupui, I understand you are eliminating the second option, but what option do you prefer ?

Reply:
I am no specialist and just read some literature today about this, but from what you're saying the current implementation seems adequate (keep `p1-p2`). Let see what others think.

Reply:
I haven't yet read up on the details of the test, so I don't yet have an informed opinion. In principle, consistency with Fisher sounds good unless:

- There is reasonable precedent for the conventions being different between the two tests
- The current `fisher_exact` really should have used the opposite convention

The most important thing is documenting the convention used, and it would be nice to provide some thoughts about why the decision went that way in code comments. Thanks for all the work on this!

Reply:
I would personally rather prefer modifying `fisher_exact` since the results from the [paper I linked](https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf) in my doc and R results are asymmetric from mine (negative when I am positive and vice-versa). But as @jb-leger told me, it is lot harder to modify code that people are currently relying into. This is why we are copying `fisher_exact`'s behavior